enhancements/monitoring/client-cert-scraping.md

```diff
@@ -0,0 +1,132 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Monitoring needs to be reliable and is the very useful when trying to debug clusters in an already degraded state.
+We want to ensure that metrics scraping can always work if the scraper can reach the target, even if the kube-apiserver
+is unavailable or unreachable.
+To do this, we will combine a local authorizer (already merged in many binaries and the rbac-proxy) and client-cert based
+authentication to have a fully local authentication and authorization path for scraper targets.
+
+## Motivation
+
+If networking (or part of networking) is down and a scraper target cannot reach the kube-apiserver to verify a token
+and a subjectaccessreview, then the metrics scraper can be rejected.
+The subjectaccessreview (authorization) is already largely addressed, but service account tokens are still used
+for scraping targets.
+Tokens require an external network call that we can avoid by using client certificates.
+Gathering metrics, especially client metrics, from partially functionally clusters helps narrow the search area between
+kube-apiserver, etcd, kubelet, and SDN considerably.
+
+In addition, this will significantly reduce the load on the kube-apiserver.
+We have observed in the CI cluster that token and subject access reviews are a significant percentage of all kube-apiserver traffic.
+
+### Goals
+
+1. Gather all possible metrics even when targets cannot reach the kube-apiserver.
+2. Avoid sending valid kube-apiserver credentials to everyone who can create a servicemonitor.
+
+### Non-Goals
+
+1. Force all servicemonitor targets to change at the same time.
+
+## Proposal
+
+1. The monitoriting operator creates a client cert/key pair valid against the kube-apiserver.
+   There is a CSR signer name specifically for this.
+2. The kube-controller-manager operator runs an approver that allows monitoring operator SA to get signed certificate for
+   The existing scraper identity.
+3. The monitoring operator puts the client cert/key pair into a secret.
+4. The monitoring operator keeps these credentials up to date (the signer is only good for 30 days or so).
+5. The metrics scraper mounts the client cert/key pair and ensures that it is hot-reloadable or builds a process suicider.
+   There is library-go code to suicide on file changes.
+6. A participating metrics scraper target, can read the in-cluster client cert CA bundle and create an authenticator
+   using it.
+   The default delegated authentication stack does this and on openshift, the configmap is exposed to system:authenticated.
+7. A servicemonitor grows a way to specify using this client cert.
```

Comment:
Not sure what you mean here. The service monitor allows to specify a client certificate/key either by file path in the Prometheus container or reference to a secret's key.

https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#tlsconfig

Reply:
> Not sure what you mean here. The service monitor allows to specify a client certificate/key either by file path in the Prometheus container or reference to a secret's key.

That appears to be a reference to a secret in the namespace containing the target.  If that were the case, then every target would have to have client cert credentials for scanning /metrics.  We want to avoid this level of permissions for every component.

Is is possible to specify a client-cert that exists in the metric scraping namespace?

Reply:
The requirement is that the secret exists in the same namespace as the service monitor. The other option today would be to use the `certFile` and `keyFile` files but it assumes that the target knows the location of these files in the Prometheus container. It's not worse than what we do now for bearer [tokens](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/07a71fcac595fd9dd9ab2a7a79e817df6cdf1675/manifests/0000_90_kube-controller-manager-operator_04_servicemonitor-controller-manager.yaml#L53) and [CA](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/07a71fcac595fd9dd9ab2a7a79e817df6cdf1675/manifests/0000_90_kube-controller-manager-operator_04_servicemonitor-controller-manager.yaml#L67) but it's deprecated and we'll probably remove it eventually.

Reply:
And I'm not saying that we can't evolve things upstream :)

Reply:
> The other option today would be to use the `certFile` and `keyFile` files but it assumes that the target knows the location of these files in the Prometheus container.

This option is more reasonable for the effort we're trying to complete here.  Having every namespace with the power to scrape all metrics is not desireable.

Reply:
> The requirement is that the secret exists in the same namespace as the service monitor. 
@simonpasquier i think referencing the secret via reference rather than file would be desirable. Do I understand correctly you think about losening the restriction in prometheus-operator to be able to reference a secret from a different namespace?

Reply:
> > The requirement is that the secret exists in the same namespace as the service monitor.
> 
> @simonpasquier i think referencing the secret via reference rather than file would be desirable. Do I understand correctly you think about losening the restriction in prometheus-operator to be able to reference a secret from a different namespace?

I think an ideal way would be to say, "send me a client cert that identifies you as the scraper" without being able to specify exactly what.  The exact mechanism can be niggled later.  For the purposes of OCP, we can even hardcode it as a default since key material never leaves the scraper pod.